# 0048
# Merging, Cleaning & Transforming Data

Here you´ll have the opportunity to code major parts of Project 4 on your own. If you need any help or inspiration, have a look at the Videos or the Jupyter Notebook with the full code. <br> <br>
Keep in mind that it´s all about __getting the right results/conclusions__. It´s not about finding the identical code. Things can be coded in many different ways. Even if you come to the same conclusions, it´s very unlikely that we have the very same code. 


In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('movies_clean.csv', parse_dates=['release_date'])

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44691 entries, 0 to 44690
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     44691 non-null  int64         
 1   title                  44691 non-null  object        
 2   tagline                20284 non-null  object        
 3   release_date           44657 non-null  datetime64[ns]
 4   genres                 42586 non-null  object        
 5   belongs_to_collection  4463 non-null   object        
 6   original_language      44681 non-null  object        
 7   budget_musd            8854 non-null   float64       
 8   revenue_musd           7385 non-null   float64       
 9   production_companies   33356 non-null  object        
 10  production_countries   38835 non-null  object        
 11  vote_count             44691 non-null  float64       
 12  vote_average           42077 non-null  float64       
 13  p

***passing the argument parse_dates=['release_date'] to the read_csv() function, which tells Pandas to parse the 'release_date' column as a datetime object. This can be useful for working with dates and times in your analysis.***

In [4]:
df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                             tagline release_date  \
0                                                NaN   1995-10-30   
1          Roll the dice and unleash the excitement!   1995-12-15   
2  Still Yelling. Still Fighting. Still Ready for...   1995-12-22   
3  Friends are the people who let you be yourself...   1995-12-22   
4  Just When His World Is Back To Normal... He's ...   1995-02-10   

                     genres           belongs_to_collection original_language  \
0   Animation|Comedy|Family            Toy Story Collection                en   
1  Adventure|Fantasy|Family                             NaN                en   
2            Romance|Comedy       Grumpy Old Men Collection                en   
3      Comedy|Drama|Romance                             NaN                en   
4                    Comedy  Father of the Bride Collection                en   

   budget_musd  revenue_musd  \
0         30.0    373.554033   
1         65.0    262.797249   
2          NaN           NaN   
3         16.0     81.452156   
4          NaN     76.578911   

                                production_companies  \
0                            Pixar Animation Studios   
1  TriStar Pictures|Teitler Film|Interscope Commu...   
2                        Warner Bros.|Lancaster Gate   
3             Twentieth Century Fox Film Corporation   
4          Sandollar Productions|Touchstone Pictures   

       production_countries  vote_count  vote_average  popularity  runtime  \
0  United States of America      5415.0           7.7   21.946943     81.0   
1  United States of America      2413.0           6.9   17.015539    104.0   
2  United States of America        92.0           6.5   11.712900    101.0   
3  United States of America        34.0           6.1    3.859495    127.0   
4  United States of America       173.0           5.7    8.387519    106.0   

                                            overview  spoken_languages  \
0  Led by Woody, Andy's toys live happily in his ...           English   
1  When siblings Judy and Peter discover an encha...  English|Français   
2  A family wedding reignites the ancient feud be...           English   
3  Cheated on, mistreated and stepped on, the wom...           English   
4  Just when George Banks has recovered from his ...           English   

                                         poster_path  
0  <img src='http://image.tmdb.org/t/p/w185//rhIR...  
1  <img src='http://image.tmdb.org/t/p/w185//vzmL...  
2  <img src='http://image.tmdb.org/t/p/w185//6ksm...  
3  <img src='http://image.tmdb.org/t/p/w185//16XO...  
4  <img src='http://image.tmdb.org/t/p/w185//e64s...

________________________

## Add DataFrame `Credits`

In [5]:
credits = pd.read_csv('credits.csv')
credits.head(5)

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

within dataframe credits , we have 3 columns named `cast`, `crew` and `id` with total rows of 45476.
these values are nested as it is shown below:

In [6]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB


In [7]:
credits['id'][0]

862

In [8]:
credits['cast'][0]

"[{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender': 2, 'id': 31, 'name': 'Tom Hanks', 'order': 0, 'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'}, {'cast_id': 15, 'character': 'Buzz Lightyear (voice)', 'credit_id': '52fe4284c3a36847f8024f99', 'gender': 2, 'id': 12898, 'name': 'Tim Allen', 'order': 1, 'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'}, {'cast_id': 16, 'character': 'Mr. Potato Head (voice)', 'credit_id': '52fe4284c3a36847f8024f9d', 'gender': 2, 'id': 7167, 'name': 'Don Rickles', 'order': 2, 'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'}, {'cast_id': 17, 'character': 'Slinky Dog (voice)', 'credit_id': '52fe4284c3a36847f8024fa1', 'gender': 2, 'id': 12899, 'name': 'Jim Varney', 'order': 3, 'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'}, {'cast_id': 18, 'character': 'Rex (voice)', 'credit_id': '52fe4284c3a36847f8024fa5', 'gender': 2, 'id': 12900, 'name': 'Wallace Shawn', 'order': 4, 'profile_path': '/oGE6JqPP2xH4t

In [9]:
credits['crew'][0]

'[{\'credit_id\': \'52fe4284c3a36847f8024f49\', \'department\': \'Directing\', \'gender\': 2, \'id\': 7879, \'job\': \'Director\', \'name\': \'John Lasseter\', \'profile_path\': \'/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f4f\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12891, \'job\': \'Screenplay\', \'name\': \'Joss Whedon\', \'profile_path\': \'/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f55\', \'department\': \'Writing\', \'gender\': 2, \'id\': 7, \'job\': \'Screenplay\', \'name\': \'Andrew Stanton\', \'profile_path\': \'/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f5b\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12892, \'job\': \'Screenplay\', \'name\': \'Joel Cohen\', \'profile_path\': \'/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f61\', \'department\': \'Writing\', \'gender\': 0, \'id\': 12893, \'job\': \'Screenplay\', \'name\': \'A

_____________________

# 0052
# Preparing the Data for Merging

### Check status of each column

In [10]:
credits['id'].value_counts()

id
141971    3
298721    2
9755      2
10991     2
99080     2
         ..
55135     1
15877     1
72272     1
1549      1
461257    1
Name: count, Length: 45432, dtype: int64

In [11]:
credits[credits.duplicated(subset=['id'], keep = False)].sort_values(by = 'id')

cast  \
25885  [{'cast_id': 12, 'character': 'The Creature', ...   
25950  [{'cast_id': 12, 'character': 'The Creature', ...   
33838  [{'cast_id': 15, 'character': 'Chuck Barris', ...   
5865   [{'cast_id': 15, 'character': 'Chuck Barris', ...   
9165   [{'cast_id': 11, 'character': 'Jef Costello', ...   
...                                                  ...   
25887  [{'cast_id': 7, 'character': 'Hollander', 'cre...   
24163  [{'cast_id': 2, 'character': 'Ebba', 'credit_i...   
45275  [{'cast_id': 2, 'character': 'Ebba', 'credit_i...   
33196  [{'cast_id': 1, 'character': 'Jenjira', 'credi...   
40051  [{'cast_id': 1, 'character': 'Jenjira', 'credi...   

                                                    crew      id  
25885  [{'credit_id': '52fe4380c3a36847f80590dd', 'de...    3057  
25950  [{'credit_id': '52fe4380c3a36847f80590dd', 'de...    3057  
33838  [{'credit_id': '52fe43e2c3a36847f80760a9', 'de...    4912  
5865   [{'credit_id': '52fe43e2c3a36847f80760b5', 'de...    4912  
9165   [{'credit_id': '52fe440ac3a36847f807ee01', 'de...    5511  
...                                                  ...     ...  
25887  [{'credit_id': '52fe4da29251416c9111ce5d', 'de...  199591  
24163  [{'credit_id': '534fd1a80e0a267eb6000e32', 'de...  265189  
45275  [{'credit_id': '534fd1a80e0a267eb6000e32', 'de...  265189  
33196  [{'credit_id': '5448c8efc3a3680fb4001582', 'de...  298721  
40051  [{'credit_id': '5739ceeac3a3683c9d000c12', 'de...  298721  

[87 rows x 3 columns]

#### Explaination on Previous Cell :
The duplicated() method in Pandas can be used to identify duplicate rows in a DataFrame based on one or more columns. In this case, you're calling the duplicated() method on the credits DataFrame, and passing the arguments subset=['id'] to specify that you want to consider only the 'id' column when identifying duplicates, and keep=False to keep all instances of duplicated rows in the DataFrame, rather than just the first or last instance.

The resulting output will be a Boolean Series where each value is True if the corresponding row is a duplicate based on the 'id' column, and False otherwise.

You're then using this Boolean Series to index the credits DataFrame using the square bracket notation ([]), which returns a new DataFrame containing only the rows that are duplicates based on the 'id' column.

Finally, you're using the sort_values() method to sort the resulting DataFrame by the 'id' column in ascending order.

The resulting output will be a new DataFrame containing only the rows that have duplicate values in the 'id' column of the credits DataFrame, sorted by the 'id' column in ascending order.

In [12]:
credits.drop_duplicates(subset='id', inplace = True)

In [13]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
Index: 45432 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45432 non-null  object
 1   crew    45432 non-null  object
 2   id      45432 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.4+ MB


#### Explaination on Previous Cell:

he drop_duplicates() method in Pandas can be used to remove duplicate rows from a DataFrame based on one or more columns. In this case, you're calling the drop_duplicates() method on the credits DataFrame, and passing the argument subset='id' to specify that you want to consider only the 'id' column when identifying duplicates.

The inplace=True parameter tells Pandas to modify the credits DataFrame in place, rather than returning a new DataFrame with the duplicates removed.

In [14]:
credits['id'].value_counts()
# output will show that for each id number we have 1 row.

id
862       1
74458     1
296206    1
107308    1
16247     1
         ..
32985     1
44399     1
10138     1
32084     1
461257    1
Name: count, Length: 45432, dtype: int64

# 0053
# Merging the Data(left Join)

In [15]:
df[~df['id'].isin(credits['id'])]

Empty DataFrame
Columns: [id, title, tagline, release_date, genres, belongs_to_collection, original_language, budget_musd, revenue_musd, production_companies, production_countries, vote_count, vote_average, popularity, runtime, overview, spoken_languages, poster_path]
Index: []

#### Explaination on Previous Cell:

The isin() method in Pandas can be used to select rows in a DataFrame based on the values in a specific column. In this case, you're calling the isin() method on the 'id' column of the credits DataFrame, and passing the resulting boolean Series to the ~ operator to select the inverse of the boolean Series. This will return a boolean Series where each value is True if the corresponding 'id' value is not present in the 'id' column of the credits DataFrame.

You're then using this boolean Series to index the df DataFrame using the square bracket notation ([]), which returns a new DataFrame containing only the rows where the 'id' value is not present in the 'id' column of the credits DataFrame.

The resulting output will be a new DataFrame containing only the rows from df where the 'id' value is not present in the 'id' column of the credits DataFrame. This can be useful for filtering out any rows in df that don't have corresponding data in the credits DataFrame.

___________________________

#### Explaination about `~` operator:

the ~ symbol in Pandas is a logical operator that is used to invert a boolean Series or DataFrame.

When you apply the ~ operator to a boolean Series or DataFrame, it will return the inverse of the boolean values. For example, if you have a boolean Series where the values are [True, False, True], applying the ~ operator will return the inverse of these values as [False, True, False].

This can be useful for filtering a DataFrame based on columns that meet certain conditions. For example, if you have a boolean Series that indicates whether each row in a DataFrame meets a certain condition, you can use the ~ operator to select the rows that do not meet that condition.

______________________

In [16]:
credits[~credits['id'].isin(df['id'])]

cast  \
189    [{'cast_id': 4, 'character': 'Himself', 'credi...   
554    [{'cast_id': 3, 'character': '', 'credit_id': ...   
682    [{'cast_id': 2, 'character': 'Eunice', 'credit...   
685    [{'cast_id': 0, 'character': 'James Bird', 'cr...   
711                                                   []   
...                                                  ...   
45320  [{'cast_id': 0, 'character': 'Mała metalowa dz...   
45323  [{'cast_id': 0, 'character': 'Krzysiek Buk', '...   
45348                                                 []   
45420  [{'cast_id': 1, 'character': 'Aprel', 'credit_...   
45452                                                 []   

                                                    crew      id  
189                                                   []   56088  
554    [{'credit_id': '52fe4e4dc3a368484e21952d', 'de...  218473  
682    [{'credit_id': '52fe475cc3a36847f81317f7', 'de...   48260  
685    [{'credit_id': '57a321c6c3a3683fa70013d8', 'de...  277270  
711                                                   []  365371  
...                                                  ...     ...  
45320  [{'credit_id': '57ee6832c3a3682a2d003ccd', 'de...  418757  
45323  [{'credit_id': '564fd6a59251414b01005a8a', 'de...  369444  
45348                                                 []  335251  
45420  [{'credit_id': '58d83051c3a368126f0567fe', 'de...  449131  
45452  [{'credit_id': '52fe4684c3a36847f81034f3', 'de...   44330  

[741 rows x 3 columns]

***the output of prvious cell has 741 rows which are thoose rows that we have cleaned from `df` dataframe***

### Merge `credits` and `df` :

In [17]:
df = df.merge(credits, how ='left', left_on = 'id', right_on='id')

The merge() method in Pandas can be used to combine two or more DataFrames into a single DataFrame based on one or more common columns. In this case, you're calling the merge() method on the df DataFrame, and passing the credits DataFrame as the second DataFrame to merge.

You're using the parameters how='left' to specify that you want to perform a left join, which will retain all rows from the df DataFrame and only include matching rows from the credits DataFrame based on the 'id' column.

You're also using the parameters left_on='id' and right_on='id' to specify that you want to merge the two DataFrames based on their 'id' column.

In [18]:
df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                             tagline release_date  \
0                                                NaN   1995-10-30   
1          Roll the dice and unleash the excitement!   1995-12-15   
2  Still Yelling. Still Fighting. Still Ready for...   1995-12-22   
3  Friends are the people who let you be yourself...   1995-12-22   
4  Just When His World Is Back To Normal... He's ...   1995-02-10   

                     genres           belongs_to_collection original_language  \
0   Animation|Comedy|Family            Toy Story Collection                en   
1  Adventure|Fantasy|Family                             NaN                en   
2            Romance|Comedy       Grumpy Old Men Collection                en   
3      Comedy|Drama|Romance                             NaN                en   
4                    Comedy  Father of the Bride Collection                en   

   budget_musd  revenue_musd  \
0         30.0    373.554033   
1         65.0    262.797249   
2          NaN           NaN   
3         16.0     81.452156   
4          NaN     76.578911   

                                production_companies  \
0                            Pixar Animation Studios   
1  TriStar Pictures|Teitler Film|Interscope Commu...   
2                        Warner Bros.|Lancaster Gate   
3             Twentieth Century Fox Film Corporation   
4          Sandollar Productions|Touchstone Pictures   

       production_countries  vote_count  vote_average  popularity  runtime  \
0  United States of America      5415.0           7.7   21.946943     81.0   
1  United States of America      2413.0           6.9   17.015539    104.0   
2  United States of America        92.0           6.5   11.712900    101.0   
3  United States of America        34.0           6.1    3.859495    127.0   
4  United States of America       173.0           5.7    8.387519    106.0   

                                            overview  spoken_languages  \
0  Led by Woody, Andy's toys live happily in his ...           English   
1  When siblings Judy and Peter discover an encha...  English|Français   
2  A family wedding reignites the ancient feud be...           English   
3  Cheated on, mistreated and stepped on, the wom...           English   
4  Just when George Banks has recovered from his ...           English   

                                         poster_path  \
0  <img src='http://image.tmdb.org/t/p/w185//rhIR...   
1  <img src='http://image.tmdb.org/t/p/w185//vzmL...   
2  <img src='http://image.tmdb.org/t/p/w185//6ksm...   
3  <img src='http://image.tmdb.org/t/p/w185//16XO...   
4  <img src='http://image.tmdb.org/t/p/w185//e64s...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...

## Join Methods:

Pandas provides four types of join methods to combine two or more DataFrames based on a common column or index. These methods are:  

**1- Inner Join:** Retains only the rows that have matching values in both DataFrames. In other words, only rows with common values in the specified join column(s) are included in the resulting DataFrame.

**2- Left Join:** Retains all rows from the left DataFrame and only includes matching rows from the right DataFrame based on the specified join column(s). If there are no matches in the right DataFrame, the resulting DataFrame will contain missing values (NaN) for those columns.  

**3- Right Join:** Retains all rows from the right DataFrame and only includes matching rows from the left DataFrame based on the specified join column(s). If there are no matches in the left DataFrame, the resulting DataFrame will contain missing values (NaN) for those columns.  

**4- Outer Join:** Retains all rows from both DataFrames and includes matching rows from both DataFrames based on the specified join column(s). If there are no matches in either DataFrame, the resulting DataFrame will contain missing values (NaN) for those columns.  

You can perform a join operation in Pandas using the merge() method, which allows you to specify the type of join to perform (how parameter), the join column(s) in each DataFrame (left_on and right_on parameters), and other parameters such as whether to retain all rows in the original DataFrames (left_index and right_index parameters).

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44691 entries, 0 to 44690
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     44691 non-null  int64         
 1   title                  44691 non-null  object        
 2   tagline                20284 non-null  object        
 3   release_date           44657 non-null  datetime64[ns]
 4   genres                 42586 non-null  object        
 5   belongs_to_collection  4463 non-null   object        
 6   original_language      44681 non-null  object        
 7   budget_musd            8854 non-null   float64       
 8   revenue_musd           7385 non-null   float64       
 9   production_companies   33356 non-null  object        
 10  production_countries   38835 non-null  object        
 11  vote_count             44691 non-null  float64       
 12  vote_average           42077 non-null  float64       
 13  p

_______________

# 0054
# Cleaning and Transforming the new "Cast" Column

In [20]:
import numpy as np
import ast

In [21]:
df['cast'][0]

"[{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender': 2, 'id': 31, 'name': 'Tom Hanks', 'order': 0, 'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'}, {'cast_id': 15, 'character': 'Buzz Lightyear (voice)', 'credit_id': '52fe4284c3a36847f8024f99', 'gender': 2, 'id': 12898, 'name': 'Tim Allen', 'order': 1, 'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'}, {'cast_id': 16, 'character': 'Mr. Potato Head (voice)', 'credit_id': '52fe4284c3a36847f8024f9d', 'gender': 2, 'id': 7167, 'name': 'Don Rickles', 'order': 2, 'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'}, {'cast_id': 17, 'character': 'Slinky Dog (voice)', 'credit_id': '52fe4284c3a36847f8024fa1', 'gender': 2, 'id': 12899, 'name': 'Jim Varney', 'order': 3, 'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'}, {'cast_id': 18, 'character': 'Rex (voice)', 'credit_id': '52fe4284c3a36847f8024fa5', 'gender': 2, 'id': 12900, 'name': 'Wallace Shawn', 'order': 4, 'profile_path': '/oGE6JqPP2xH4t

In [22]:
df['cast'] = df['cast'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str)else np.nan)

In [23]:
pd.DataFrame(df['cast'][0])

cast_id                character                 credit_id  gender  \
0        14            Woody (voice)  52fe4284c3a36847f8024f95       2   
1        15   Buzz Lightyear (voice)  52fe4284c3a36847f8024f99       2   
2        16  Mr. Potato Head (voice)  52fe4284c3a36847f8024f9d       2   
3        17       Slinky Dog (voice)  52fe4284c3a36847f8024fa1       2   
4        18              Rex (voice)  52fe4284c3a36847f8024fa5       2   
5        19             Hamm (voice)  52fe4284c3a36847f8024fa9       2   
6        20          Bo Peep (voice)  52fe4284c3a36847f8024fad       1   
7        26             Andy (voice)  52fe4284c3a36847f8024fc1       0   
8        22              Sid (voice)  52fe4284c3a36847f8024fb1       2   
9        23       Mrs. Davis (voice)  52fe4284c3a36847f8024fb5       1   
10       24         Sergeant (voice)  52fe4284c3a36847f8024fb9       2   
11       25           Hannah (voice)  52fe4284c3a36847f8024fbd       1   
12       27     TV Announcer (voice)  52fe4284c3a36847f8024fc5       2   

         id               name  order                      profile_path  
0        31          Tom Hanks      0  /pQFoyx7rp09CJTAb932F2g8Nlho.jpg  
1     12898          Tim Allen      1  /uX2xVf6pMmPepxnvFWyBtjexzgY.jpg  
2      7167        Don Rickles      2  /h5BcaDMPRVLHLDzbQavec4xfSdt.jpg  
3     12899         Jim Varney      3  /eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg  
4     12900      Wallace Shawn      4  /oGE6JqPP2xH4tNORKNqxbNPYi7u.jpg  
5      7907  John Ratzenberger      5  /yGechiKWL6TJDfVE2KPSJYqdMsY.jpg  
6      8873        Annie Potts      6  /eryXT84RL41jHSJcMy4kS3u9y6w.jpg  
7   1116442        John Morris      7  /vYGyvK4LzeaUCoNSHtsuqJUY15M.jpg  
8     12901    Erik von Detten      8  /twnF1ZaJ1FUNUuo6xLXwcxjayBE.jpg  
9     12133     Laurie Metcalf      9  /unMMIT60eoBM2sN2nyR7EZ2BvvD.jpg  
10     8655       R. Lee Ermey     10  /r8GBqFBjypLUP9VVqDqfZ7wYbSs.jpg  
11    12903      Sarah Freeman     11                              None  
12    37221      Penn Jillette     12  /zmAaXUdx12NRsssgHbk1T31j2x9.jpg

using the apply() method to apply a lambda function to the 'cast' column. The lambda function uses the ast.literal_eval() function to convert the 'cast' column from a string representation of a list to an actual list object.

You're then using the isinstance() function to check if the 'cast' column is a string, and returning np.nan if it is not.

Finally, you're using the pd.DataFrame() function to create a new DataFrame from the first row of the 'cast' column.

The resulting output will be a new DataFrame containing the data from the first row of the 'cast' column, with each actor's name and character name in separate columns.

In [24]:
df['cast_size'] =df['cast'].apply(lambda x: len(x))

In [25]:
df['cast_size'].value_counts(dropna = False).head(20)

cast_size
10    2770
8     2729
7     2710
6     2649
5     2637
9     2557
15    2444
4     2406
11    2283
0     2189
12    2043
13    1799
14    1578
16    1350
1     1302
17    1103
3     1087
2      882
18     811
19     785
Name: count, dtype: int64

In [26]:
df['cast'] = df['cast'].apply(lambda x: "|".join(i['name'] for i in x) if isinstance(x, list) else np.nan)

In [27]:
df['cast']

0        Tom Hanks|Tim Allen|Don Rickles|Jim Varney|Wal...
1        Robin Williams|Jonathan Hyde|Kirsten Dunst|Bra...
2        Walter Matthau|Jack Lemmon|Ann-Margret|Sophia ...
3        Whitney Houston|Angela Bassett|Loretta Devine|...
4        Steve Martin|Diane Keaton|Martin Short|Kimberl...
                               ...                        
44686              Leila Hatami|Kourosh Tahami|Elham Korda
44687    Angel Aquino|Perry Dizon|Hazel Orencio|Joel To...
44688    Erika Eleniak|Adam Baldwin|Julie du Page|James...
44689    Iwan Mosschuchin|Nathalie Lissenko|Pavel Pavlo...
44690                                                     
Name: cast, Length: 44691, dtype: object

In [28]:
df.head(3)

id             title                                            tagline  \
0    862         Toy Story                                                NaN   
1   8844           Jumanji          Roll the dice and unleash the excitement!   
2  15602  Grumpier Old Men  Still Yelling. Still Fighting. Still Ready for...   

  release_date                    genres      belongs_to_collection  \
0   1995-10-30   Animation|Comedy|Family       Toy Story Collection   
1   1995-12-15  Adventure|Fantasy|Family                        NaN   
2   1995-12-22            Romance|Comedy  Grumpy Old Men Collection   

  original_language  budget_musd  revenue_musd  \
0                en         30.0    373.554033   
1                en         65.0    262.797249   
2                en          NaN           NaN   

                                production_companies  ... vote_count  \
0                            Pixar Animation Studios  ...     5415.0   
1  TriStar Pictures|Teitler Film|Interscope Commu...  ...     2413.0   
2                        Warner Bros.|Lancaster Gate  ...       92.0   

   vote_average  popularity  runtime  \
0           7.7   21.946943     81.0   
1           6.9   17.015539    104.0   
2           6.5   11.712900    101.0   

                                            overview  spoken_languages  \
0  Led by Woody, Andy's toys live happily in his ...           English   
1  When siblings Judy and Peter discover an encha...  English|Français   
2  A family wedding reignites the ancient feud be...           English   

                                         poster_path  \
0  <img src='http://image.tmdb.org/t/p/w185//rhIR...   
1  <img src='http://image.tmdb.org/t/p/w185//vzmL...   
2  <img src='http://image.tmdb.org/t/p/w185//6ksm...   

                                                cast  \
0  Tom Hanks|Tim Allen|Don Rickles|Jim Varney|Wal...   
1  Robin Williams|Jonathan Hyde|Kirsten Dunst|Bra...   
2  Walter Matthau|Jack Lemmon|Ann-Margret|Sophia ...   

                                                crew cast_size  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...        13  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...        26  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...         7  

[3 rows x 21 columns]

In [29]:
df['cast'].value_counts(dropna = False).head(10)

cast
                      2189
Georges Méliès          24
Louis Theroux           15
Mel Blanc               12
Jimmy Carr               9
George Carlin            8
Louis C.K.               8
David Attenborough       8
Werner Herzog            8
Patton Oswalt            6
Name: count, dtype: int64

In [30]:
df['cast'].replace("", np.nan, inplace = True)

In [31]:
df['cast'].value_counts(dropna = False).head(10)


cast
NaN                   2189
Georges Méliès          24
Louis Theroux           15
Mel Blanc               12
Jimmy Carr               9
George Carlin            8
Louis C.K.               8
David Attenborough       8
Werner Herzog            8
Patton Oswalt            6
Name: count, dtype: int64

________________________________

# 0055
# Cleaning And Transforming the New `Crew` Column

In [32]:
df['crew'][0]

'[{\'credit_id\': \'52fe4284c3a36847f8024f49\', \'department\': \'Directing\', \'gender\': 2, \'id\': 7879, \'job\': \'Director\', \'name\': \'John Lasseter\', \'profile_path\': \'/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f4f\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12891, \'job\': \'Screenplay\', \'name\': \'Joss Whedon\', \'profile_path\': \'/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f55\', \'department\': \'Writing\', \'gender\': 2, \'id\': 7, \'job\': \'Screenplay\', \'name\': \'Andrew Stanton\', \'profile_path\': \'/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f5b\', \'department\': \'Writing\', \'gender\': 2, \'id\': 12892, \'job\': \'Screenplay\', \'name\': \'Joel Cohen\', \'profile_path\': \'/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg\'}, {\'credit_id\': \'52fe4284c3a36847f8024f61\', \'department\': \'Writing\', \'gender\': 0, \'id\': 12893, \'job\': \'Screenplay\', \'name\': \'A

In [33]:
df['crew'] = df['crew'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [34]:
df['crew'][0]

[{'credit_id': '52fe4284c3a36847f8024f49',
  'department': 'Directing',
  'gender': 2,
  'id': 7879,
  'job': 'Director',
  'name': 'John Lasseter',
  'profile_path': '/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f4f',
  'department': 'Writing',
  'gender': 2,
  'id': 12891,
  'job': 'Screenplay',
  'name': 'Joss Whedon',
  'profile_path': '/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f55',
  'department': 'Writing',
  'gender': 2,
  'id': 7,
  'job': 'Screenplay',
  'name': 'Andrew Stanton',
  'profile_path': '/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f5b',
  'department': 'Writing',
  'gender': 2,
  'id': 12892,
  'job': 'Screenplay',
  'name': 'Joel Cohen',
  'profile_path': '/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg'},
 {'credit_id': '52fe4284c3a36847f8024f61',
  'department': 'Writing',
  'gender': 0,
  'id': 12893,
  'job': 'Screenplay',
  'name': 'Alec Sokolow',
  'profile_path': '/v79vlRYi94BZUQnkkyzn

This conversion is done to transform the string representation of a list into an actual list object.
utilizing the isinstance() function to check if the value in the 'crew' column is a string. If it is, the lambda function applies ast.literal_eval() to convert it into a list. If the value is not a string, it sets it as np.nan.

By performing this operation, you're effectively converting the 'crew' column from a string representation to a list representation, making it easier to work with the data

In [35]:
pd.DataFrame(df['crew'][0])

credit_id  department  gender       id  \
0    52fe4284c3a36847f8024f49   Directing       2     7879   
1    52fe4284c3a36847f8024f4f     Writing       2    12891   
2    52fe4284c3a36847f8024f55     Writing       2        7   
3    52fe4284c3a36847f8024f5b     Writing       2    12892   
4    52fe4284c3a36847f8024f61     Writing       0    12893   
..                        ...         ...     ...      ...   
101  589216c19251412dc2009cb9       Sound       1  1748724   
102  589216ccc3a3680973009274       Sound       0  1748725   
103  589216d79251412dc8009aa0       Sound       0     8067   
104  589216e49251412dcd009a4f  Production       1     7902   
105  589216f39251412dc2009cf3  Production       0    84493   

                        job            name                      profile_path  
0                  Director   John Lasseter  /7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg  
1                Screenplay     Joss Whedon  /dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg  
2                Screenplay  Andrew Stanton  /pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg  
3                Screenplay      Joel Cohen  /dAubAiZcvKFbboWlj7oXOkZnTSu.jpg  
4                Screenplay    Alec Sokolow  /v79vlRYi94BZUQnkkyznbGUZLjT.jpg  
..                      ...             ...                               ...  
101  Assistant Sound Editor   Susan Sanford                              None  
102  Assistant Sound Editor   Susan Popovic                              None  
103  Assistant Sound Editor    Dan Engstrom                              None  
104      Casting Consultant    Ruth Lambert                              None  
105       ADR Voice Casting  Mickie McGowan  /k7TjJBfINsg8vLQxJwos6XObAD6.jpg  

[106 rows x 7 columns]

By using df['crew'][0], you're selecting the 'crew' column from the df DataFrame and indexing it with [0] to access the first row of the column.

Next, you're using the pd.DataFrame() function to create a new DataFrame from the first row of the 'crew' column. This will create a DataFrame where each element of the first row is placed in a separate column.

In [36]:
df['crew_size'] = df['crew'].apply(lambda x: len(x))

The lambda function lambda x: len(x) is applied to each value in the 'crew' column, where x represents each element of the 'crew' list. The len() function is then used to determine the length of the 'crew' list, which gives you the number of crew members for each movie.

By using the apply() method in conjunction with the lambda function, you're able to iterate over each row in the 'crew' column and calculate the size of the 'crew' list for that particular row.

In [37]:
df['crew_size'].value_counts(dropna = False).head(20)

crew_size
2     6197
3     4951
1     4866
4     3064
5     2240
7     1968
10    1928
6     1867
8     1864
9     1686
11    1555
12    1388
13    1181
14    1017
15     854
16     746
17     633
0      620
18     541
19     420
Name: count, dtype: int64

In [38]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
        return np.nan

#### Explaination On `get_director()` :
 iterating over the elements of the input list x using a for loop. Each element represents a crew member.

Within the loop, you're checking if the 'job' attribute of the crew member is equal to 'Director'. If it is, you're returning the value of the 'name' attribute, which represents the director's name.

If none of the crew members have a 'job' attribute equal to 'Director', you're returning np.nan (NumPy's representation of missing values) to indicate that the director's name is not available.

It's important to note that the return np.nan statement should be outside the for loop, as it should only be executed if none of the crew members have a 'job' attribute equal to 'Director'. In your current code, the return statement is within the loop, which may cause the function to return prematurely after checking only the first crew member.

In [39]:
df['director'] = df['crew'].apply(get_director)

In [40]:
df['director']

0           John Lasseter
1                     NaN
2           Howard Deutch
3         Forest Whitaker
4                     NaN
               ...       
44686    Hamid Nematollah
44687            Lav Diaz
44688      Mark L. Lester
44689    Yakov Protazanov
44690       Daisy Asquith
Name: director, Length: 44691, dtype: object

In [41]:
df['director'].value_counts(dropna = False).head(20)

director
NaN                 19017
None                  620
Georges Méliès         46
John Ford              45
Alfred Hitchcock       38
Charlie Chaplin        37
Werner Herzog          37
Robert Wise            35
Takashi Miike          33
Frank Capra            33
Clint Eastwood         32
Michael Curtiz         31
Buster Keaton          30
Otto Preminger         27
Sidney Lumet           27
Roger Corman           26
William Beaudine       26
Woody Allen            25
Fritz Lang             25
Ingmar Bergman         25
Name: count, dtype: int64

__________________

## Check the DataFrame

In [42]:
df.head(5)

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                             tagline release_date  \
0                                                NaN   1995-10-30   
1          Roll the dice and unleash the excitement!   1995-12-15   
2  Still Yelling. Still Fighting. Still Ready for...   1995-12-22   
3  Friends are the people who let you be yourself...   1995-12-22   
4  Just When His World Is Back To Normal... He's ...   1995-02-10   

                     genres           belongs_to_collection original_language  \
0   Animation|Comedy|Family            Toy Story Collection                en   
1  Adventure|Fantasy|Family                             NaN                en   
2            Romance|Comedy       Grumpy Old Men Collection                en   
3      Comedy|Drama|Romance                             NaN                en   
4                    Comedy  Father of the Bride Collection                en   

   budget_musd  revenue_musd  \
0         30.0    373.554033   
1         65.0    262.797249   
2          NaN           NaN   
3         16.0     81.452156   
4          NaN     76.578911   

                                production_companies  ... popularity  runtime  \
0                            Pixar Animation Studios  ...  21.946943     81.0   
1  TriStar Pictures|Teitler Film|Interscope Commu...  ...  17.015539    104.0   
2                        Warner Bros.|Lancaster Gate  ...  11.712900    101.0   
3             Twentieth Century Fox Film Corporation  ...   3.859495    127.0   
4          Sandollar Productions|Touchstone Pictures  ...   8.387519    106.0   

                                            overview  spoken_languages  \
0  Led by Woody, Andy's toys live happily in his ...           English   
1  When siblings Judy and Peter discover an encha...  English|Français   
2  A family wedding reignites the ancient feud be...           English   
3  Cheated on, mistreated and stepped on, the wom...           English   
4  Just when George Banks has recovered from his ...           English   

                                         poster_path  \
0  <img src='http://image.tmdb.org/t/p/w185//rhIR...   
1  <img src='http://image.tmdb.org/t/p/w185//vzmL...   
2  <img src='http://image.tmdb.org/t/p/w185//6ksm...   
3  <img src='http://image.tmdb.org/t/p/w185//16XO...   
4  <img src='http://image.tmdb.org/t/p/w185//e64s...   

                                                cast  \
0  Tom Hanks|Tim Allen|Don Rickles|Jim Varney|Wal...   
1  Robin Williams|Jonathan Hyde|Kirsten Dunst|Bra...   
2  Walter Matthau|Jack Lemmon|Ann-Margret|Sophia ...   
3  Whitney Houston|Angela Bassett|Loretta Devine|...   
4  Steve Martin|Diane Keaton|Martin Short|Kimberl...   

                                                crew cast_size crew_size  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...        13       106   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...        26        16   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...         7         4   
3  [{'credit_id': '52fe44779251416c91011acb', 'de...        10        10   
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...        12         7   

          director  
0    John Lasseter  
1              NaN  
2    Howard Deutch  
3  Forest Whitaker  
4              NaN  

[5 rows x 23 columns]

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44691 entries, 0 to 44690
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     44691 non-null  int64         
 1   title                  44691 non-null  object        
 2   tagline                20284 non-null  object        
 3   release_date           44657 non-null  datetime64[ns]
 4   genres                 42586 non-null  object        
 5   belongs_to_collection  4463 non-null   object        
 6   original_language      44681 non-null  object        
 7   budget_musd            8854 non-null   float64       
 8   revenue_musd           7385 non-null   float64       
 9   production_companies   33356 non-null  object        
 10  production_countries   38835 non-null  object        
 11  vote_count             44691 non-null  float64       
 12  vote_average           42077 non-null  float64       
 13  p

In [45]:
df.drop(columns = ['crew'], inplace = True)

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44691 entries, 0 to 44690
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     44691 non-null  int64         
 1   title                  44691 non-null  object        
 2   tagline                20284 non-null  object        
 3   release_date           44657 non-null  datetime64[ns]
 4   genres                 42586 non-null  object        
 5   belongs_to_collection  4463 non-null   object        
 6   original_language      44681 non-null  object        
 7   budget_musd            8854 non-null   float64       
 8   revenue_musd           7385 non-null   float64       
 9   production_companies   33356 non-null  object        
 10  production_countries   38835 non-null  object        
 11  vote_count             44691 non-null  float64       
 12  vote_average           42077 non-null  float64       
 13  p

_______________

# Save df

In [47]:
df.to_csv('movies_compl.csv', index = False)

***specifying index = False to exclude the index column from being saved in the CSV file. This is useful if you don't want the index values to be included as a separate column in the CSV file.***